sync15-adapter/src/state.rs

```diff
@@ -46,3 +73,568 @@ impl GlobalState {
+                self.client.put_crypto_keys(&new_keys)?;
+
+                Ok(InitialWithLiveToken)
+            }
+        }
+    }
+
+    /// Runs through the state machine to the ready state.
+    pub fn to_ready(&mut self) -> error::Result<()> {
+        let mut state = InitialWithLiveToken;
+        loop {
+            match &state {
+                Ready => {
+                    self.sequence.push(mem::discriminant(&state));
+                    break;
+                }
+                // If we already started over once before, we're likely in a
+                // cycle, and should try again later. Like the iOS state
+                // machine, other cycles aren't a problem; we'll cycle through
+                // earlier states if we need to reupload `meta/global` or
+                // `crypto/keys`.
+                FreshStartRequired { .. } if self.sequence.contains(&mem::discriminant(&state)) => {
+                    return Err(ErrorKind::SetupStateCycleError.into());
+                }
+                _ => {
+                    self.sequence.push(mem::discriminant(&state));
+                    state = self.advance(state)?;
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+enum SetupState {
+    InitialWithLiveToken,
+    InitialWithLiveTokenAndInfo,
+    NeedsFreshMetaGlobal {
+        previous_global: Option<BsoRecord<MetaGlobalRecord>>,
+    },
+    HasMetaGlobal,
+    ResolveMetaGlobal {
+        previous_global: Option<BsoRecord<MetaGlobalRecord>>,
+        new_global: BsoRecord<MetaGlobalRecord>,
+    },
+    NeedsFreshCryptoKeys,
+    Ready,
+    FreshStartRequired {
+        previous_global: Option<BsoRecord<MetaGlobalRecord>>,
+    },
+}
+
+/// Whether we should skip fetching `meta/global` or `crypto/keys` from the
+/// server because our locally cached copy is up-to-date, fetch a fresh copy
+/// from the server, or invalidate our locally cached state, then upload a
+/// fresh record.
+enum FetchAction {
+    Skip,
+    Fetch,
+    InvalidateThenUpload,
+}
+
+impl FetchAction {
+    fn from_modified(
+        local: Option<&ServerTimestamp>,
+        remote: Option<&ServerTimestamp>,
+    ) -> FetchAction {
+        match (local, remote) {
+            (Some(local), Some(remote)) => {
+                if *local >= *remote {
+                    FetchAction::Skip
+                } else {
+                    FetchAction::Fetch
+                }
+            }
+            (Some(_), None) => FetchAction::InvalidateThenUpload,
+            _ => FetchAction::Fetch,
+        }
+    }
+}
+
+/// Flags an engine for enablement or disablement.
+pub enum EngineStateChange {
+    ResetAll { except: HashSet<String> },
+    Enable(String),
+    Disable(String),
+    Reset(String),
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use reqwest;
+
+    use bso_record::{BsoRecord, EncryptedBso, EncryptedPayload};
+
+    struct InMemoryClient {
```

Comment:
I think this kind of test client is really useful, but I think it would be more useful to build more of the actual functionality.  That is, rather than storing `BsoRecord<...>`, store a server timestamp and a `MetaGlobal` record and _produce_ the `BsoRecord` on demand, like the service would.  Same with `crypto/keys`: store the inputs and the timestamp, and act like the service really does.